### PR TITLE
Distinguish canceling a route pop from bubbling the pop up to the next level

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by jason-simmons** [comment]
@abarth

File: packages/flutter/lib/src/widgets/routes.dart [hunk 1]
```diff
@@ -385,11 +387,23 @@ class _ModalScopeStatus extends InheritedWidget {
...
+  /// Delegate this to the next level of navigation.
+  bubble,
+}
+
 /// Signature for a callback that verifies that it's OK to call [Navigator.pop].
 ///
 /// Used by [Form.onWillPop], [ModalRoute.addScopedWillPopCallback], and
 /// [ModalRoute.removeScopedWillPopCallback].
-typedef Future<bool> WillPopCallback();
+typedef Future<RoutePopDisposition> WillPopCallback();
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/routes.dart]: I wonder if we should leave the callback returning a Boolean.  It's not obvious to me that returning "bubble" means "keep calling the rest of the callbacks".

**Comment by abarth** [comment]
LGTM with the callback switched back to a bool

File: packages/flutter/lib/src/widgets/routes.dart (hunk 1 above)
Comment (jason-simmons) [inline packages/flutter/lib/src/widgets/routes.dart]: done

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -72,7 +73,8 @@ abstract class Route<T> {
   /// See also:
   ///
   /// * [Form], which provides an `onWillPop` callback that uses this mechanism.
-  Future<bool> willPop() async => !isFirst;
+  Future<RoutePopDisposition> willPop() async => isFirst ?
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: please don't use => with multiline expressions

**Comment by Hixie** [comment]
This needs a test.

File: packages/flutter/lib/src/widgets/routes.dart [hunk 3]
```diff
@@ -385,6 +387,18 @@ class _ModalScopeStatus extends InheritedWidget {
   }
 }
 
+/// Indicates whether the current route should be popped.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/routes.dart]: Please add a breadcrumb to willPop.

File: packages/flutter/lib/src/widgets/routes.dart [hunk 4]
```diff
@@ -385,6 +387,18 @@ class _ModalScopeStatus extends InheritedWidget {
...
+enum RoutePopDisposition {
+  /// Pop the route.
+  pop,
+
+  /// Do not pop the route.
+  doNotPop,
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/routes.dart]: The docs for pop and doNotPop don't follow the style guide (both could be written by someone who knows nothing except the name of the enum and the name of the value). Please add more colour to these explaining more about what they mean, e.g.: what their implications are, how you might use them, code samples, methods that use these values, how they interact with Android or with nested Navigators, etc.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/routes.dart]: (or leave them blank so we'll catch them later)

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 2 above)
Comment (jason-simmons) [inline packages/flutter/lib/src/widgets/navigator.dart]: done

**Comment by jason-simmons** [comment]
Followed up on feedback and added a test

File: examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart [hunk 5]
```diff
@@ -75,12 +75,12 @@ class TextFormFieldDemoState extends State<TextFormFieldDemo> {
     return null;
   }
 
-  Future<bool> _warnUserAboutInvalidData() {
+  Future<bool> _warnUserAboutInvalidData() async {
     final FormState form = _formKey.currentState;
-    if (!_formWasEdited || form.validate())
+    if (form == null || !_formWasEdited || form.validate())
       return new Future<bool>.value(true);
```
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/material/text_form_field_demo.dart]: just return true now that you're async.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**